File: src/main/java/me/andre111/d20server/util/RollFormatter.java
package me.andre111.d20server.util;

import me.andre111.d20common.model.profile.Profile;
import me.andre111.d20common.scripting.expression.Result;
import me.andre111.d20server.service.ChatService;

public class RollFormatter {
	//TODO: this should not include the player specific part
	public static String formatDiceRoll(Profile profile, String rollExpression, boolean showPublic, Result result, Exception e) {
		// build "header"
		StringBuilder sb = new StringBuilder();
		sb.append("<p class=\"chat-sender\">");
		sb.append(ChatService.escape(profile.getName()));
		if(!showPublic) {
			sb.append(" (to GM)");
		}
		sb.append(": ");
		
		sb.append("</p>");
		
		sb.append("<span class=\"hoverable\">");
		sb.append("<p class=\"chat-info\">");
		sb.append("rolling ...");
		sb.append("</p>");
		sb.append("<div class=\"onhover\">");
		sb.append(ChatService.escape(normalizeExpression(rollExpression)));
		sb.append("</div>");
		sb.append("</span>");

		// result
		sb.append("<p class=\"chat-message\">");
		if(result != null) {
			sb.append(result.getString());
			sb.append("<br>");
			sb.append(" = ");
			appendResultValue(sb, result);
		} else {
			sb.append(" = ?");
		}
		sb.append("</p>");
		
		// potential error message
		if(e != null) {
			sb.append("<p class=\"chat-info\">");
			sb.append("( ");
			sb.append(ChatService.escape(e.getMessage()));
			sb.append(" )");
			sb.append("</p>");
		}
		
		return sb.toString();
	}
	
	public static String formatInlineDiceRoll(String rollExpression, Result result, String e) {
		StringBuilder sb = new StringBuilder();
		// color format
		String color = "#000000";
		if(result != null) {
			if(result.hadCriticalFailure() && result.hadCriticalSuccess()) {
				color = "#0000FF";
			} else if(result.hadCriticalFailure()) {
				color = "#FF0000";
			} else if(result.hadCriticalSuccess()) {
				color = "#008800";
			}
		}
		
		// total value
		sb.append("<span style=\"color:"+color+"\" class=\"chat-dice-inline hoverable\">");
		if(result != null) {
			appendResultValue(sb, result);
		} else {
			sb.append("?");
		}
		// show full result on hover
		sb.append("<div class=\"onhover\">");
		sb.append(result.getString());
		sb.append("</div>");
		sb.append("</span>");
		
		// show expression on hover
		sb.append("<span class=\"hoverable\">");
		sb.append("*");
		sb.append("<div class=\"onhover\">");
		sb.append(ChatService.escape(normalizeExpression(rollExpression)));
		sb.append("</div>");
		sb.append("</span>");
		
		return sb.toString();
	}
	
	private static void appendResultValue(StringBuilder sb, Result result) {
		if(Math.round(result.getValue()) == result.getValue()) {
			sb.append(Integer.toString((int) result.getValue()));
		} else {
			sb.append(Double.toString(result.getValue()));
		}
	}
	
	private static String normalizeExpression(String expression) {
		expression = expression.replace("+", " + ");
		expression = expression.replace("-", " - ");
		expression = expression.replace("*", " * ");
		expression = expression.replace("/", " / ");
		expression = expression.replace("(", "( ");
		expression = expression.replace(",", ", ");
		expression = expression.replace(")", " )");
		expression = expression.replaceAll(" +", " ");
		return expression;
	}
}

File: core/server/util/roll-formatter.js
import { I18N } from '../../common/util/i18n.js';
import { ChatService } from '../service/chat-service.js';

export class RollFormatter {
    static formatDiceRoll(profile, rollExpression, result, appendix = '') {
        // build 'header'
        var text = '<div class="chat-sender">';
        text = text + ChatService.escape(profile.getUsername()) + appendix + ': ';
        text = text + '</div>';

        text = text + '<span class="hoverable">';
        text = text + '<div class="chat-info">' + I18N.get('chat.rolling', 'rolling...') + '</div>';
        text = text + '<div class="onhover">' + ChatService.escape(RollFormatter.normalizeExpression(rollExpression)) + '</div>';
        text = text + '</span>';

        // result
        text = text + '<div class="chat-message">';
        if (result) {
            text = text + result.expr + '<br>';
            text = text + ' = ' + RollFormatter.getResultValue(result);
        } else {
            text = text + ' = ?';
        }
        text = text + '</div>';

        return text;
    }

    static formatInlineDiceRoll(rollExpression, result, diceRolls, error) {
        // check crits
        var hadCriticalFailure = false;
        var hadCriticalSuccess = false;
        for (const roll of diceRolls) {
            if (roll.cf) hadCriticalFailure = true;
            if (roll.cs) hadCriticalSuccess = true;
        }

        // color format
        var color = '#000000';
        if (hadCriticalFailure && hadCriticalSuccess) {
            color = '#0000FF';
        } else if (hadCriticalFailure) {
            color = '#FF0000';
        } else if (hadCriticalSuccess) {
            color = '#008800';
        }

        // total value
        var text = `<span style="color:${color}" class="chat-dice-inline hoverable">`;
        if (result) {
            text = text + this.getResultValue(result);
        } else {
            text = text + '?';
        }
        // show full result on hover
        text = text + `<div class="onhover${!result && error ? ' chat-error' : ''}">` + (result ? result.expr : (error ? error : '')) + '</div>';
        text = text + '</span>';

        // show expression on hover
        text = text + '<span class="hoverable">';
        text = text + '*';
        text = text + '<div class="onhover">' + ChatService.escape(RollFormatter.normalizeExpression(rollExpression)) + '</div>';
        text = text + '</span>';

        return text;
    }

    static getResultValue(result) {
        if (Math.trunc(result.value) == result.value) {
            return String(Math.trunc(result.value));
        } else {
            return String(result.value);
        }
    }

    static normalizeExpression(expression) {
        // prevents double escaping by reverting already escaped characters
        expression = ChatService.unescape(expression);

        expression = expression.replace(/\+/g, ' + ');
        expression = expression.replace(/-/g, ' - ');
        expression = expression.replace(/\*/g, ' * ');
        expression = expression.replace(/\//g, ' / ');
        expression = expression.replace(/\(/g, '( ');
        expression = expression.replace(/,/g, ', ');
        expression = expression.replace(/\)/g, ' )');
        expression = expression.replace(/ +/g, ' ');
        return expression;
    }
}
